app/discovery/provider/docker.go

```diff
@@ -272,3 +257,79 @@ func contains(e string, s []string) bool {
 	}
 	return false
 }
+
+type dockerClient struct {
+	client  http.Client
+	network string // network for IP selection
+}
+
+// NewDockerClient constructs docker client for given host and network
+func NewDockerClient(host, network string) DockerClient {
+	var schemaRegex = regexp.MustCompile("^(?:([a-z0-9]+)://)?(.*)$")
+	parts := schemaRegex.FindStringSubmatch(host)
+	proto, addr := parts[1], parts[2]
+	log.Printf("[DEBUG] Configuring docker client to talk to %s via %s", addr, proto)
+
+	client := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial(proto, addr)
+			},
+		},
+		Timeout: time.Second * 5,
+	}
+
+	return &dockerClient{client, network}
+}
+
+func (d *dockerClient) ListContainers() ([]containerInfo, error) {
+
+	const dockerAPIVersion = "v1.41"
+
+	resp, err := d.client.Get(fmt.Sprintf("http://localhost/%s/containers/json", dockerAPIVersion))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed connection to docker socket")
+	}
+
+	defer resp.Body.Close()
+
+	response := []struct {
+		containerInfo
+		Created         int64
+		NetworkSettings struct {
+			Networks map[string]struct {
+				IPAddress string
+			}
+		}
+		Names        []string
+		ExposedPorts []struct{ PrivatePort int } `json:"Ports"`
+	}{}
+
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return nil, errors.Wrap(err, "failed to parse response from docker daemon")
+	}
+
+	containers := make([]containerInfo, len(response))
+
+	for i, c := range response {
+		// fill remaining fields
+		c.TS = time.Unix(c.Created, 0)
+
+		for k, v := range c.NetworkSettings.Networks {
+			if d.network == "" || k == d.network { // match on network name if defined
+				c.IP = v.IPAddress
```

Comment:
unless I missed it, I don't think you don't handle "no networks" error. Not even sure if this is a thing but probably returning an error will be nice

Reply:
In case there are no networks defined the IP will remain empty and container will later will be ignored in `(d *Docker) listContainers()`. Should I check this anyway?

Reply:
it this case already handled - up to you.

Reply:
I think it is wrong for docker client to be responsible for this kind of validation. I already filter networks based on dockerClient config which feels less than ideal.